Add recommended labels into operands&OLM workloads

erkanerol opened the pull request:
With this PR, we are adding recommended labels[1] into all operands we manage with hco-operator.

````
app.kubernetes.io/managed-by=hco-operator
app.kubernetes.io/version=<version-of-hco-operator>
app.kubernetes.io/component=<component-of-operand-like-compute-or-storage>
app.kubernetes.io/part-of=hyperconverged-cluster
````

We are keeping this as it is in order to not break any existing mechanism.
```
app=kubevirt-hyperconverged
```

***OLM PART***
This PR adds the labels into the workloads we deploy with OLM as well.  See the changes in
- deploy/index-image/kubevirt-hyperconverged/1.4.0/kubevirt-hyperconverged-operator.v1.4.0.clusterserviceversion.yaml
- deploy/olm-catalog/kubevirt-hyperconverged/1.4.0/kubevirt-hyperconverged-operator.v1.4.0.clusterserviceversion.yaml

***operator.yaml***
This PR adds the labels into the workloads we deploy with operator.yaml. Since we don't know who is going to deploy with this file, we didn't use "managed-by" label in this manifest file. 

***Testing***

- Unit tests in hyperconverged_controller_test.go and e2e tests in check_labels.sh: checks all objects under .status.relatedObjects of HCO and verifies that all of them have these labels.



**Release note**:
```release-note
Adds some of the recommended labels* into workloads

* https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/ 
```

1. https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

erkanerol commented:
/retest

tiraboschi commented:
/test hco-e2e-upgrade-prev-aws
/test hco-e2e-upgrade-prev-azure

tiraboschi commented:
/test hco-e2e-upgrade-prev-aws
/test hco-e2e-upgrade-prev-azure

tiraboschi commented:
/test hco-e2e-upgrade-prev-aws

nunnatsa reviewed (commented):
Not sure I like the long `if` conditions in the `updateCr` methods. 

Each `if` checks different field in, and then, within the if block, but there is no relation between the field.

so instead:

```golang
if !reflect.DeepEqual(require.A, exists.A) || !reflect.DeepEqual(require.B, exists.B) {
    handle A
    handle B
}
```

I think it will be clrearer to do:
```golang
if !reflect.DeepEqual(require.A, exists.A) {
    handle A
}

if !reflect.DeepEqual(require.B, exists.B) {
    handle B
}
```

erkanerol commented:
/retest

zcahana commented on pkg/components/components.go:
```diff
@@ -116,7 +116,10 @@ func GetDeploymentSpecOperator(namespace, image, imagePullPolicy, conversionCont
 				Labels: map[string]string{
-					"name": hcoName,
+					"name":                    hcoName,
+					hcoutil.AppVersionLabel:   hcoKvIoVersion,
```
Here you add the labels directly to the podSpec - shouldn't this be added to the Deployment's labels too?
Same comment applies both to the operator and webhook deployments.

erkanerol replied:
I discussed this with @tiraboschi  and explained that OLM doesn't allow us to set labels for deployments. It takes only DeploymentSpec.

zcahana replied:
Seems like this capability was recently added in [operator-framework/api#46](https://github.com/operator-framework/api/pull/46) (although [its parent issue](https://github.com/operator-framework/operator-lifecycle-manager/issues/1522) is still open).

erkanerol replied:
I added the labels into the StrategyDeploymentSpec objects as well. OLM is not currently using them but my understanding is that it is going to use.

zcahana commented on pkg/util/consts.go:
```diff
@@ -38,3 +50,15 @@ const (
+	StorageComponent                 = "storage"
+	ImportComponent                  = "import"
+	NetworkComponent                 = "network"
+	MonitorComponent                 = "monitor"
```
I think "monitoring" or maybe "metrics" could be a better name than "monitor".

erkanerol replied:
After the discussions in chat, I updated this part as "monitoring". Also changed the names of enums like the ServiceType naming of k8s https://github.com/kubernetes/kubernetes/blob/ddf3eb5a1877338da806c2be15b53f456b6342c9/pkg/apis/core/types.go#L3446

zcahana commented on pkg/util/consts.go:
```diff
@@ -38,3 +50,15 @@ const (
+const (
+	ComputeComponent    AppComponent = "compute"
+	StorageComponent                 = "storage"
+	ImportComponent                  = "import"
```
Don't know if "import" merits a component name of its own (i.e., if its significant enough for the user?). Maybe this can fit as part of "compute" component? Or maybe "migration" is a better name?

erkanerol replied:
It was suggested by @fabiand  at the beginning of the discussions.

zcahana commented on pkg/util/labels.go:
```diff
@@ -0,0 +1,12 @@
+
+func DeepCopyLabels(src, tgt *metav1.ObjectMeta) {
+	if src.Labels != nil {
+		tgt.Labels = make(map[string]string, len(src.Labels))
```
It would be safer to "append" to the target labels instead of overwriting - i.e. check if already exists before creating...

erkanerol replied:
We are checking the deep equality for labels since we are so opinionated about the labels of our operands. This was a result of a discussion with @tiraboschi

erkanerol replied:
we talked this with @zcahana  and we are ok with it.

zcahana commented on pkg/util/labels.go:
```diff
@@ -0,0 +1,12 @@
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+func DeepCopyLabels(src, tgt *metav1.ObjectMeta) {
+	if src.Labels != nil {
```
I think it's nicer to:
```go
if src.Labels == nil {
	return
}
// rest of code, unindented
```

erkanerol replied:
It sounds better! Just fixed.

zcahana commented on tools/csv-merger/csv-merger.go:
```diff
@@ -456,6 +499,20 @@ func main() {
 }
 
+func overwriteWithStandardLabels(labels map[string]string, version string, component hcoutil.AppComponent) {
+	appendedLabels := map[string]string{
```
What's the use of the extra map + for loop? (except that it indents a bit more nicely)
Why not just this 4-liner:
```go
labels[hcoutil.AppManagedByLabel] = "olm"
labels[hcoutil.AppVersionLabel] = version
labels[hcoutil.AppPartOfLabel] = hcoutil.HyperConvergedCluster
labels[hcoutil.AppComponentLabel] = string(component)
```

erkanerol replied:
Good catch!

zcahana commented on pkg/apis/hco/v1alpha1/hyperconverged_resources.go:
```diff
@@ -19,23 +21,27 @@ func (r *HyperConverged) getNamespace(defaultNamespace string, opts []string) st
-		hcoutil.AppLabel: hcoName,
+		hcoutil.AppLabel:          hcoName,
+		hcoutil.AppManagedByLabel: hcoutil.OperatorName,
+		hcoutil.AppVersionLabel:   os.Getenv(hcoutil.HcoKvIoVersionName),
```
Also, here you're setting the `HCO_KV_IO_VERSION` version to all operands, although in the CSV you're setting the operand specific version (e.g., CNAO 0.44.0) - need to decide on a consistent approach. If the operand specific version is used, then probably the application name label should reflect that (e.g., `hcoutil.AppLabel: "cnao"`).

zcahana commented on pkg/apis/hco/v1alpha1/hyperconverged_resources.go:
```diff
@@ -19,23 +21,27 @@ func (r *HyperConverged) getNamespace(defaultNamespace string, opts []string) st
 
 	return map[string]string{
-		hcoutil.AppLabel: hcoName,
+		hcoutil.AppLabel:          hcoName,
```
What about the `app.kubernetes.io/instance` instance label, should we be adding that as well?

erkanerol replied:
We discussed which labels we need to implement and possible key/value pairs. We didn't invest on this "instance" label at the moment.

zcahana commented:
A few more general comments:
1. I think this PR can have a release note.
2. I know you already tested this, but would it be easy to add an automated test that verifies that all `relatedObjects` have these labels?
3. Note regarding reconciliation of labels - if we want to ensure that the labels are always reconciled, even when changed by the user, then we need to modify the [predicate](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/7098156c11a82a8ff96b79621f9daf01b664cce2/pkg/controller/hyperconverged/hyperconverged_controller.go#L114) we're using such that it'll also accept label changes (currently it only admits changes to spec and/or annotations).

erkanerol commented on pkg/util/consts.go:
```diff
@@ -25,6 +25,18 @@ const (
 	APIVersion             = APIVersionGroup + "/" + APIVersionBeta
+	// Recommended labels by Kubernetes. See
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
+	AppLabelPrefix    = "app.kubernetes.io"
```
IT makes sense!! I just updated as you recommended.

zcahana commented on pkg/apis/hco/v1alpha1/hyperconverged_resources.go:
```diff
@@ -19,23 +21,27 @@ func (r *HyperConverged) getNamespace(defaultNamespace string, opts []string) st
-		hcoutil.AppLabel: hcoName,
+		hcoutil.AppLabel:          hcoName,
+		hcoutil.AppManagedByLabel: hcoutil.OperatorName,
+		hcoutil.AppVersionLabel:   os.Getenv(hcoutil.HcoKvIoVersionName),
```
Correcting myself as per @yuvalturg comment, there aren't syscalls involved in reading the Environment variable - these are served by the runtime from memory, and hence the cost of `Getenv()` multiple times is reasonable. Nevertheless this might be somewhat cleaner to read them once, store, and serve from a state variable.

erkanerol replied:
fixed.

erkanerol commented:
> A few more general comments:
> 
>     1. I think this PR can have a release note.
> 
>     2. I know you already tested this, but would it be easy to add an automated test that verifies that all `relatedObjects` have these labels?
> 
>     3. Note regarding reconciliation of labels - if we want to ensure that the labels are always reconciled, even when changed by the user, then we need to modify the [predicate](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/7098156c11a82a8ff96b79621f9daf01b664cce2/pkg/controller/hyperconverged/hyperconverged_controller.go#L114) we're using such that it'll also accept label changes (currently it only admits changes to spec and/or annotations).

1. I added a release note.
2. I added both unit tests and e2e tests. 
3. After checking the code base, we noticed that we don't use predicate for secondary resources so there is no need a change.

erkanerol commented:
/retest

erkanerol commented:
/test hco-e2e-upgrade-aws
/test hco-e2e-upgrade-prev-aws

zcahana commented:
Looks good @erkanerol!
/lgtm

erkanerol commented:
/test hco-e2e-image-index-gcp

zcahana commented:
@erkanerol no need to `/test hco-e2e-image-index-gcp` since `hco-e2e-image-index-aws` and `hco-e2e-image-index-azure` have passed, so our friend `kubevirt-bot` will make sure to `/override hco-e2e-image-index-gcp`.

tiraboschi commented:
/approve
